core/src/main/java/io/grpc/internal/SharingExecutorService.java

```diff
@@ -0,0 +1,114 @@
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wraps an existing executor, that keeps track of only the tasks it has added.
+ */
+public final class SharingExecutorService extends AbstractExecutorService {
+  private final Map<CleanupRunnable, Future<?>> activeTasks =
+      new HashMap<CleanupRunnable, Future<?>>();
+
+  private final ExecutorService delegate;
+  private boolean shutdown;
+  private boolean terminated;
+
+  public SharingExecutorService(ExecutorService delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public synchronized void shutdown() {
+    shutdown = true;
+    if (activeTasks.isEmpty()) {
+      terminated = true;
+      notifyAll();
+    }
+  }
+
+  @Override
+  public synchronized List<Runnable> shutdownNow() {
+    shutdown();
+    List<Runnable> neverRan = new ArrayList<Runnable>(activeTasks.size());
+    for (Entry<CleanupRunnable, Future<?>> entry : activeTasks.entrySet()) {
+      // Since the underlying executor may have started running this, this is best effort.
+      boolean success = entry.getValue().cancel(true);
```

Comment:
I made it interrupting because ThreadPoolExecutor does it as well.  shutdownNow seems to be a more forceful intent than shutdown.  Given that this method is likely not going to be used by use (we use regular shutdown at all the callsites this class is intended for) it doesn't feel too dangerous.  I mostly implemented this method for correctness.